test/run-functional-tests

```diff
@@ -0,0 +1,419 @@
+	if [ "$?" -ne 0 ];then
+		fatal "$DOCKER_SERVICE is not active"
+	fi
+	# check if docker cgroup driver is systemd
+	docker info 2> /dev/null|grep -q "Cgroup Driver: systemd"
+	if [ "$?" -ne 0 ];then
+		warning "docker cgroup driver suggest to be \"systemd\""
+	fi
+
+	[ "$1" == "cleanup" ] && return
+	# check if ha network already exists
+	for network in ${HA_NETWORK_ARRAY[@]};do
+		docker network ls|grep -q "$network"
+		if [ "$?" -eq 0 ];then
+			fatal "HA specific network \"$network\" already exists"
+		fi
+	done
+}
+
+
+get_test_case_array() {
+	test -d $BEHAVE_CASE_DIR || fatal "Cannot find '$BEHAVE_CASE_DIR'"
+	ls $BEHAVE_CASE_DIR|grep "\.feature"|grep -Ev "$BEHAVE_CASE_EXCLUDE"
+}
+
+
+echo_test_cases() {
+	case_arry=`get_test_case_array`
+	echo "Index|File Name|Description"
+	index=1
+	 for f in ${case_arry[@]};do
+		desc=`awk -F: '/Feature/{print $2}' $BEHAVE_CASE_DIR/$f`
+		printf "%3s    %-40s %-60s\n" $index $f "$desc"
+		index=$(($index+1))
+	done
+	printf "%3s    %-40s  %-60s\n" $index "regression test" "Original regression test"
+}
+
+
+usage_and_exit() {
+	prog=`basename $0`
+	cat <<END
+Usage: $prog [OPTIONS]|[TESTCASE INDEX]
+$prog is a tool for developers to setup the cluster in containers to run functional tests.
+The container image is based on Tumbleweed with preinstalled packages of the cluster stack include pacemaker/corosync/crmsh and many others.
+Users can make the code change under crmsh.git including test cases. This tool will pick up the code change and "make install" to all running containers.
+
+OPTIONS:
+  -h, --help		Show this help message and exit
+  -l            	List existing functional test cases and exit
+  -n NUM        	Only setup a cluster with NUM nodes(containers)
+  -x			Don't config corosync on containers(with -n option)
+  -d            	Cleanup the cluster containers
+
+EXAMPLES:
+To launch 2 nodes with the running cluster with the very basic corosync.conf
+# crmsh.git/test/run-functional-tests -n 2
+
+To launch 2 nodes without the cluster stack running to play with "crm cluster init/join"
+# crmsh.git/run-functional-tests -n 2 -x
+
+To list the existing test cases. Users could add his own new test cases.
+# crmsh.git/test/run-functional-tests -l
+
+To run a single or a number of functional test cases
+# crmsh.git/test/run-functional-tests 1
+# crmsh.git/test/run-functional-tests 1 2 3
+
+To clean up the all containers which are generated by this tool
+# crmsh.git/test/run-functional-tests -d
+END
+	exit 1
+}
+
+
+docker_exec() {
+	name=$1
+	cmd=$2
+	docker exec -t $name /bin/sh -c "$cmd"
+}
+
+
+deploy_ha_node() {
+	node_name=$1
+	project_path=$(dirname $(dirname `realpath $0`))
+	docker_options="-d --name $node_name -h $node_name --privileged --shm-size 1g -v $project_path:/app"
+        build_cmd="cd /app;./autogen.sh && ./configure --prefix /usr && make install && make install-crmconfDATA prefix="
+
+	info "Deploying \"$node_name\"..."
+	docker run $docker_options $DOCKER_IMAGE &> /dev/null
```

Comment:
In a fresh rebooted machine, docker containers already deployed by "run-functional-test" might fail to start. "docker start" is expected instead of "docker run". So, additional detection and handling are expected.

Reply:
Changed

run `./autogen.sh && ./configure --prefix /usr` on host before create container

Reply:
Changed as `docker run --restart always`, and run `crm cluster enable` before running `crm cluster start`
So after rebooted, the containers will started, too